app/org/sagebionetworks/bridge/services/UploadArchiveService.java

```diff
@@ -1,38 +1,40 @@
 package org.sagebionetworks.bridge.services;
 
-import java.io.ByteArrayInputStream;
-import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.security.cert.CertificateEncodingException;
-import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.ExecutionException;
-import java.util.zip.ZipEntry;
-import java.util.zip.ZipInputStream;
-import java.util.zip.ZipOutputStream;
 
 import javax.annotation.Nonnull;
 
-import com.google.common.base.Strings;
-import com.google.common.cache.LoadingCache;
-import com.google.common.util.concurrent.UncheckedExecutionException;
-import org.apache.commons.io.IOUtils;
 import org.bouncycastle.cms.CMSException;
-
+import org.sagebionetworks.bridge.config.BridgeConfigFactory;
 import org.sagebionetworks.bridge.crypto.CmsEncryptor;
 import org.sagebionetworks.bridge.exceptions.BadRequestException;
 import org.sagebionetworks.bridge.exceptions.BridgeServiceException;
+import org.sagebionetworks.bridge.util.DuplicateZipEntryException;
+import org.sagebionetworks.bridge.util.ZipOverflowException;
+import org.sagebionetworks.bridge.util.Zipper;
 import org.sagebionetworks.bridge.validators.Validate;
-
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
 
+import com.google.common.base.Strings;
+import com.google.common.cache.LoadingCache;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+
 /**
  * Archives messages and then encrypts the archive using CMS.
  * Conversely also decrypts the archive and unpacks it.
  */
 @Component
 public class UploadArchiveService {
+
+    /** Max number of uncompressed bytes per zip entry. */
+    private final Zipper zipper = new Zipper(
```

Comment:
This probably doesn't matter since UploadArchiveService is a singleton, but can we also make this static?